javalib/src/main/scala/java/util/Collections.scala

```diff
@@ -0,0 +1,1131 @@
+      case _                  => new ReverseComparator
+    }
+  }
+
+  def enumeration[T](c: Collection[T]): Enumeration[T] = {
+    val it = c.iterator
+    new Enumeration[T] {
+      override def hasMoreElements: Boolean =
+        it.hasNext
+
+      override def nextElement(): T =
+        it.next()
+    }
+  }
+
+  def list[T](e: Enumeration[T]): ArrayList[T] = {
+    val arrayList = new ArrayList[T]
+    e.foreach(arrayList += _)
+    arrayList
+  }
+
+  def frequency(c: Collection[_], o: AnyRef): Int =
+    c.count(_ === o)
+
+  def disjoint(c1: Collection[_], c2: Collection[_]): Boolean = {
+    if (c1.size < c2.size)
+      !c1.exists(elem => c2.contains(elem))
+    else
+      !c2.exists(elem => c1.contains(elem))
+  }
+
+  def addAll[T](c: Collection[_ >: T], elements: Array[AnyRef]): Boolean =
+    c.addAll(asJavaCollection(elements.asInstanceOf[Array[T]]))
+
+  def newSetFromMap[E](map: Map[E, Boolean]): Set[E] = {
+    if (map.nonEmpty)
+      throw new IllegalArgumentException
+
+    new WrappedSet[E, Set[E]] {
+      override protected val inner: Set[E] =
+        map.keySet
+
+      override def add(e: E): Boolean =
+        map.put(e, true) === null
+
+      override def addAll(c: Collection[_ <: E]): Boolean =
+        c.foldLeft(false)((prev, elem) => map.put(elem, true) === null || prev)
+    }
+  }
+
+  @inline
+  private def modulo(a: Int, b: Int): Int = ((a % b) + b) % b
+
+  @inline
+  private def makeComparator[T <: jl.Comparable[T]]: Comparator[T] = {
+    new Comparator[T] {
+      final def compare(o1: T, o2: T): Int = o1.compareTo(o2)
+    }
+  }
+
+  @inline
+  private implicit def comparatorToOrdering[E](cmp: Comparator[E]): Ordering[E] = {
+    new Ordering[E] {
+      final def compare(x: E, y: E): Int = cmp.compare(x, y)
+    }
+  }
+
+  @inline
+  private def binarySearchCore[E](list: List[E], compareToKey: E => Int): Int = {
+    def notFound(insertionPoint: Int): Int = {
+      -insertionPoint - 1
+    }
+    list match {
+      case _: RandomAccess =>
+        @tailrec
+        def binarySearch(lo: Int, hi: Int): Int = {
+          if (lo < hi) {
+            val mid = lo + (hi - lo) / 2
+            val cmp = compareToKey(list(mid))
+            if (cmp === 0) mid
+            else if (cmp > 0) binarySearch(lo, mid)
+            else
+              binarySearch(mid + 1, hi)
+          } else {
+            notFound(lo)
+          }
+        }
+        binarySearch(0, list.size)
+      case _ =>
+        @tailrec
+        def linearSearch(iter: Iterator[E], index: Int): Int = {
+          if (iter.hasNext) {
+            val cmp = compareToKey(iter.next())
+            if (cmp == 0)
+              index
+            else if (cmp > 0)
+              notFound(index)
+            else
+              linearSearch(iter, index + 1)
+          } else {
+            notFound(list.size)
+          }
+        }
+        linearSearch(list.iterator, 0)
```

Comment:
This implement does not satisfy the complexity constraints of the JavaDoc:

> If the specified list does not implement the RandomAccess interface and is large, this method will do an iterator-based binary search that performs O(n) link traversals and O(log n) element comparisons.

The trick here is to use a `listIterator()`, and basically do a binary-search algorithm. And then, when `mid` changes, do something like a "go-to" in the list implemented with something like this:
```scala
def goToIndexUpwards(iter: ListIterator[_], index: Int): Unit = {
  for (i <- 0 until index - iter.nextIndex)
    iter.next()
}

def goToIndexDownwards(iter: ListIterator[_], index: Int): Unit = {
  for (i <- 0 until iter.nextIndex - index)
    iter.previous()
  }
}
```
Using `goToIndexUpwards` in the case `cmp < 0` and `goToIndexDownwards` int he case `cmp > 0`.